### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 1]
```diff
@@ -542,93 +543,73 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
...
+          s"Intermediate field ${newCheckedFields.quoted} does not exist.")
+      }
+    case _ =>
+      val fieldName = if (checkedFields.isEmpty) {
+        "The first parameter"
+      } else {
+        s"Intermediate field ${checkedFields.quoted}"
+      }
+      TypeCheckResult.TypeCheckFailure(
+        s"$fieldName must be struct type, got: ${currentStruct.catalogString}")
```
Comment (fqaiser94) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:587]: this code does not check if intermediate fields are struct type. As a result, we  get `java.lang.ClassCastException` thrown by line 593 if any intermediate field is not a struct type. Personally, I would prefer an AnalysisException with a relatively descriptive message like I have in my PR. So I wanted to ask, is this change intentional?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 2]
```diff
@@ -542,93 +543,73 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
...
-        case (name, i) if resolver(name, attribute) => Seq(Literal(attribute),
-          toCreateNamedStruct(GetStructField(struct, i, Some(attribute)), nextAttributes))
+      CreateNamedStruct(existingFields.zipWithIndex.flatMap {
+        case (name, i) if resolver(name, fieldPath.head) =>
+          val innerStruct = GetStructField(struct, i, Some(fieldPath.head))
+          Seq(Literal(fieldPath.head), toCreateNamedStruct(innerStruct, fieldPath.tail))
         case (name, i) => Seq(Literal(name), GetStructField(struct, i, Some(name)))
       })
     }
+    If(IsNull(struct), Literal(null, expr.dataType), expr)
```
Comment (fqaiser94) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:613]: I think it would be better if we do something like: 
```
    if (struct.nullable) {
      If(IsNull(struct), Literal(null, expr.dataType), expr)
    } else {
      expr
    }
```

Right now, with the code in this PR, WithField will always be nullable even if the original struct was not nullable. This change would fix that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 3]
```diff
@@ -542,93 +543,73 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
...
-  """)
-// scalastyle:on line.size.limit
-case class WithField(children: Seq[Expression]) extends Unevaluable {
-  private lazy val Seq(structExpr, nestedAttributeExpr, valExpr) = children
-  private lazy val nestedAttribute = nestedAttributeExpr.eval().toString
-  private lazy val attributes = UnresolvedAttribute.parseAttributeName(nestedAttribute)
-  lazy val toCreateNamedStruct: Expression = toCreateNamedStruct(structExpr, attributes)
+case class WithField(struct: Expression, fieldPath: Seq[String], value: Expression)
+  extends Unevaluable {
+  assert(fieldPath.nonEmpty)
```
Comment (fqaiser94) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:549]: Would it not be better to do this in `checkInputDataTypes` function? So we get an AnalysisException with a descriptive message?

**Review by fqaiser94 — COMMENTED** [review]
Hey @cloud-fan , thanks for doing this. 
I'm more than happy to adopt most of these changes. 
I see however that this does not pass any of the unit tests. Am I correct in thinking this PR was just your way of sharing your ideas? And that I should fix up the tests/code as necessary after merging it? Please let me know.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:549]: I do `assert` here as this should be a bug. Users can't construct an empty `fieldPath` with public APIs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:613]: It's not a big deal as the optimizer will turn `IsNull` to false literal if the expression is not nullable. You can see from `ScalaReflection` that we also add `IsNull` without checking nullability, to leave optimization work to the optimizer.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:587]: sorry, made a typo, I've pushed a new commit to fix it.

**Comment by cloud-fan** [comment]
Yea I didn't fix all the tests. I just tried my idea and ran some local examples. Please help to fix the tests after merging it, thanks!
